File: tests/unit/ChecklistItemTests.cpp
#include <QtTest>
#include <QSignalSpy>
#include "domain/ChecklistItem.h"

class ChecklistItemTests : public QObject
{
    Q_OBJECT

private slots:
    void testDefaults();
    void testTextProperty();
    void testDoneProperty();
    void testToggle();
    void testSignalTextChanged();
    void testSignalDoneChanged();
};

void ChecklistItemTests::testDefaults()
{
    ChecklistItem item;
    QVERIFY(!item.id().isEmpty());
    QCOMPARE(item.text(), QString(""));
    QCOMPARE(item.done(), false);
}

void ChecklistItemTests::testTextProperty()
{
    ChecklistItem item("Test item");
    QCOMPARE(item.text(), QString("Test item"));

    item.setText("Updated");
    QCOMPARE(item.text(), QString("Updated"));
}

void ChecklistItemTests::testDoneProperty()
{
    ChecklistItem item("Test");
    QCOMPARE(item.done(), false);

    item.setDone(true);
    QCOMPARE(item.done(), true);
}

void ChecklistItemTests::testToggle()
{
    ChecklistItem item("Test");
    QCOMPARE(item.done(), false);

    item.toggle();
    QCOMPARE(item.done(), true);

    item.toggle();
    QCOMPARE(item.done(), false);
}

void ChecklistItemTests::testSignalTextChanged()
{
    ChecklistItem item("Original");
    QSignalSpy spy(&item, &ChecklistItem::textChanged);

    item.setText("Changed");
    QCOMPARE(spy.count(), 1);

    item.setText("Changed");
    QCOMPARE(spy.count(), 1); // No signal for same value
}

void ChecklistItemTests::testSignalDoneChanged()
{
    ChecklistItem item;
    QSignalSpy spy(&item, &ChecklistItem::doneChanged);

    item.setDone(true);
    QCOMPARE(spy.count(), 1);

    item.setDone(true);
    QCOMPARE(spy.count(), 1); // No signal for same value
}

QTEST_MAIN(ChecklistItemTests)
#include "ChecklistItemTests.moc"

File: csharp/tests/Checklist.Core.Tests/ChecklistItemTests.cs
using Checklist.Domain;
using Xunit;

namespace Checklist.Core.Tests;

public class ChecklistItemTests
{
    [Fact]
    public void Defaults()
    {
        var item = new ChecklistItem();
        Assert.False(string.IsNullOrEmpty(item.Id));
        Assert.Equal("", item.Text);
        Assert.False(item.Done);
    }

    [Fact]
    public void TextProperty()
    {
        var item = new ChecklistItem("Test item");
        Assert.Equal("Test item", item.Text);

        item.Text = "Updated";
        Assert.Equal("Updated", item.Text);
    }

    [Fact]
    public void DoneProperty()
    {
        var item = new ChecklistItem("Test");
        Assert.False(item.Done);

        item.Done = true;
        Assert.True(item.Done);
    }

    [Fact]
    public void Toggle()
    {
        var item = new ChecklistItem("Test");
        Assert.False(item.Done);

        item.Toggle();
        Assert.True(item.Done);

        item.Toggle();
        Assert.False(item.Done);
    }

    [Fact]
    public void SignalTextChanged()
    {
        var item = new ChecklistItem("Original");
        var count = 0;
        item.TextChanged += () => count++;

        item.Text = "Changed";
        Assert.Equal(1, count);

        item.Text = "Changed";
        Assert.Equal(1, count);
    }

    [Fact]
    public void SignalDoneChanged()
    {
        var item = new ChecklistItem();
        var count = 0;
        item.DoneChanged += () => count++;

        item.Done = true;
        Assert.Equal(1, count);

        item.Done = true;
        Assert.Equal(1, count);
    }
}